### PR TITLE
Xgcm docs edit

### LINKED ISSUE #318: Interpolate CMIP6 data into reanalysis grid
Hello! Thanks for this great package, I’ve been able to use xgcm to transform data variables into density coordinates. I was also hoping to use it to interpolate CMIP6 data into [reanalysis data depth levels](https://github.com/jdldeauna/GitHub-Issues/blob/master/Interpolate%20CMIP6%20data%20to%20reanalysis%20grid.ipynb), but I’m confused as to how to assign coordinates to do this appropriately. I also wonder how data such as World Ocean Atlas (which has less info available compared to CMIP6 data) can be used for interpolation. I would appreciate any insights. Thank you so much!

### PR DESCRIPTION
**Pull request by jdldeauna**
- [X] Closes #318 
 - [X] Revise ordering of examples (depth before salinity to temperature space interpolation)
 - [X] Simplified depth interpolation example and updated to use CNRM model data

### CODE REVIEW

File: doc/transform.ipynb [hunk 1]
```diff
@@ -2335,8 +2335,14 @@
    "cell_type": "markdown",
```
Comment (jbusecke) [inline doc/transform.ipynb:2335]: <p>I think "regular depth grid" might be better? Also, by regular do you mean the distances are regularly spaced?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xgcm/xgcm/pull/322/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

Comment (jdldeauna) [reply doc/transform.ipynb:2335]: <p>Q1: Yes, "regular depth grid" would be better.</p><p> </p><p>Q2: I did mean a regularly spaced vertical grid, but maybe "regular z-level vertical grid" would be a better description?</p>

Comment (dcherian) [reply doc/transform.ipynb:2335]: I like "regularly spaced vertical grid" or "uniformly spaced..."

Comment (jbusecke) [reply doc/transform.ipynb:2335]: <p>+1 for uniformly spaced</p>

Comment (jbusecke) [inline doc/transform.ipynb:2335]: <p>I think the depth-&gt;depth interpolation should actually come first. Its the simpler application in my opinion.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xgcm/xgcm/pull/322/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (jbusecke) [inline doc/transform.ipynb:2335]: <p>Maybe we should even take a step back and call it "Interpolate to different depth coordinates"?</p><p> </p><p>Reading this, I think that this example is great, but very complex. It involves a lot of other dependencies, that might distract the reader from the grid.transform part.</p><p> </p><p>My suggestion:</p><ul><li>Keep this notebook separately, and lets turn it into an example in <a href="https://github.com/xgcm/xgcm-examples" target="_blank">xgcm-examples</a>. To me that is a place to show of 'full stack' examples.</li></ul><p> </p><ul><li>For this part here, I would just take the dataset that we already use for the other examples (GODAS) and convert it to a uniform depth (you can simply specify that by <code>np.linspace</code> or <code>np.arange</code>). The point to make here is that we can use <code>grid.transform</code> so simply interpolate on different depth values.</li></ul><p> </p><p>What do you think?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xgcm/xgcm/pull/322/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='54'/>

Comment (jbusecke) [inline doc/transform.ipynb:2335]: <p><code>CFSR_thetao = xr.open_dataset(URL,engine='netcdf4')</code>  do you need to specify <code>engine</code>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xgcm/xgcm/pull/322/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='57'/>

Comment (jdldeauna) [reply doc/transform.ipynb:2335]: <p>How's this for the final version: "Interpolating to uniformly spaced vertical grid"</p><p> </p><p> </p>

Comment (jdldeauna) [reply doc/transform.ipynb:2335]: <blockquote>Keep this notebook separately, and lets turn it into an example in&nbsp;<a href="https://github.com/xgcm/xgcm-examples" target="_blank">xgcm-examples</a>. To me that is a place to show of 'full stack' examples.</blockquote><p>	Sounds good!</p><p> </p><blockquote>For this part here, I would just take the dataset that we already use for the other examples (GODAS) and convert it to a uniform depth (you can simply specify that by&nbsp;<code>np.linspace</code>&nbsp;or&nbsp;<code>np.arange</code>). The point to make here is that we can use&nbsp;<code>grid.transform</code>&nbsp;so simply interpolate on different depth values.</blockquote><p> </p><p>I can work on an example, but I wanted to confirm that this is a correct way of assigning values:</p><p> </p><pre class="ql-syntax" spellcheck="false">grid = Grid(cmip6_thetao, coords={'Z': {'center': 'lev', 'outer': 'lev_bounds'},
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'X': {'center': 'lon'},
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'Y': {'center': 'lat'}},
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;periodic=False
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;)
</pre><p> </p><p> </p><p> </p>

Comment (jdldeauna) [reply doc/transform.ipynb:2335]: <p>I think I was working with a previous version that required me to specify engine. Here's how I download the data now:</p><p> </p><pre class="ql-syntax" spellcheck="false">cfsr_thetao_nc = netCDF4.Dataset(URL)
cfsr_thetao = xr.open_dataset(xr.backends.NetCDF4DataStore(cfsr_thetao_nc))
</pre>

**Comment by jbusecke** [comment]
I am having trouble reviewing changes in the notebook with reviewnb. I am getting this:
image

@andersy005 @dcherian, do you know what could have happened here?

Ill review now on the docs page directly.

**Comment by jbusecke** [comment]
This looks very nice! Thank you for all the work on this!

I have one comment on this line:
```
level_outer_data = np.pad(ds.lev.values, [1, 0])
```
This will actually not give you the proper cell edges. This will give you the cell center position (`lev`) and add a zero at the top. What you actually want is depth values, that 'bracket' each of the `lev` values. Thankfully these are already given by `lev_bounds` in the dataset. I noticed that in my example `lev_bounds` gets dropped, but you can avoid that with a fairly simple change:
```python
# from this
thetao = ddict['CMIP.CNRM-CERFACS.CNRM-ESM2-1.historical.r1i1p1f2.Omon.thetao.gn.gs://cmip6/CMIP6/CMIP/CNRM-CERFACS/CNRM-ESM2-1/historical/r1i1p1f2/Omon/thetao/gn/v20181206/.nan.20181206'].thetao
so = ddict['CMIP.CNRM-CERFACS.CNRM-ESM2-1.historical.r1i1p1f2.Omon.so.gn.gs://cmip6/CMIP6/CMIP/CNRM-CERFACS/CNRM-ESM2-1/historical/r1i1p1f2/Omon/so/gn/v20181206/.nan.20181206'].so
vo = ddict['CMIP.CNRM-CERFACS.CNRM-ESM2-1.historical.r1i1p1f2.Omon.vo.gn.gs://cmip6/CMIP6/CMIP/CNRM-CERFACS/CNRM-ESM2-1/historical/r1i1p1f2/Omon/vo/gn/v20181206/.nan.20181206'].vo

vo = vo.rename({'y':'y_c', 'lon':'lon_v', 'lat':'lat_v', 'bounds_lon':'bounds_lon_v', 'bounds_lat':'bounds_lat_v'})
ds = xr.merge([thetao,so,vo], compat='override')
ds


# to this
thetao = ddict['CMIP.CNRM-CERFACS.CNRM-ESM2-1.historical.r1i1p1f2.Omon.thetao.gn.gs://cmip6/CMIP6/CMIP/CNRM-CERFACS/CNRM-ESM2-1/historical/r1i1p1f2/Omon/thetao/gn/v20181206/.nan.20181206']
so = ddict['CMIP.CNRM-CERFACS.CNRM-ESM2-1.historical.r1i1p1f2.Omon.so.gn.gs://cmip6/CMIP6/CMIP/CNRM-CERFACS/CNRM-ESM2-1/historical/r1i1p1f2/Omon/so/gn/v20181206/.nan.20181206']
vo = ddict['CMIP.CNRM-CERFACS.CNRM-ESM2-1.historical.r1i1p1f2.Omon.vo.gn.gs://cmip6/CMIP6/CMIP/CNRM-CERFACS/CNRM-ESM2-1/historical/r1i1p1f2/Omon/vo/gn/v20181206/.nan.20181206']

vo = vo.rename({'y':'y_c', 'lon':'lon_v', 'lat':'lat_v', 'bounds_lon':'bounds_lon_v', 'bounds_lat':'bounds_lat_v'})
ds = xr.merge([thetao,so,vo], compat='override')
ds
```
Note the removed `.so` (or other variables) in the new version.

Then you can reconstruct your outer level coordinates using cf-xarray:
```python
import cf_xarray
level_outer_data = cf_xarray.bounds_to_vertices(ds.lev_bounds, 'axis_nbounds').load().data
```

To illustrate the difference I made a little plot:
```python
level_outer_data_old = np.pad(ds.lev.values, [1, 0])

import matplotlib.pyplot as plt
plt.plot(np.ones(4)*0, level_outer_data_old[0:4], marker='*', label='old bounds')
plt.plot(np.ones(4)*0.2,level_outer_data[0:4], marker='d', label='new bounds bounds')
plt.plot(np.ones(4)*0.1,ds.lev.data[0:4], marker='o', label='cell centers')
plt.xlim(-0.1, 0.4)
plt.grid()
plt.legend()
plt.ylabel('Depth')
```
image

You can see that except the first point the old outer coordinate (blue) is equal to the actual cell centers (green). Meanwhile the values constructed from `lev_bounds`(orange) is properly surrounding the cell centers. 

A few additional small nitpicks:

- I think it would be nice to have a further level of headings (e.g. "Depth to Depth" and "Depth to Potential Temperature"). You can just add this in the markdown cells with `#### Depth to Depth` and `#### Depth to Potential Temperature`.

- Could you add an additional `#` in front of Conservative Interpolation? I just noticed that is not indented properly in the site 
image

With these small changes, this is ready to merge in my eyes!

**Comment by jbusecke** [comment]
I will also switch the status of the PR to "Ready for review".

**Comment by jbusecke** [comment]
> I am having trouble reviewing changes in the notebook with reviewnb. I am getting this:
> image
> 
> @andersy005 @dcherian, do you know what could have happened here?
> 
> Ill review now on the docs page directly.

Hmm nevermind, now it works...not sure what was going on there.

**Comment by jdldeauna** [comment]
Thanks the comments @jbusecke ! I have one more comment before I push the changes. This test:

```python
# Warning: this step takes a long time to compute. We will only compare the first time value
xr.testing.assert_allclose(v_transport_theta_manual.isel(time=0), v_transport_theta.isel(time=0))

```
fails with the realistic data example with the following message:

```python
AssertionError: Left and right DataArray objects are not close
Differing dimensions:
    (y_c: 294, x: 362, theta: 37) != (y_c: 294, x: 362, theta_on_v: 37)
Coordinates only on the left object:
  * theta       (theta) float64 -1.5 -0.5 0.5 1.5 2.5 ... 31.5 32.5 33.5 34.5
Coordinates only on the right object:
  * theta_on_v  (theta_on_v) float64 -1.5 -0.5 0.5 1.5 ... 31.5 32.5 33.5 34.5
```

I'm not sure how to proceed because the dimensions and coordinates look similar. Should I continue to comment it out or delete the lines for now?

**Comment by jbusecke** [comment]
This is a naming issue. 

```
(y_c: 294, x: 362, **theta**: 37) != (y_c: 294, x: 362, **theta_on_v**: 37)
the dimensions between ** ** have the same values, but different names!
```

The new vertical naming will be determined by the name of the dataarray you provide for `target_data`. You have two options: 
1) rename the datarray before transforming
2) Renaming the dimension for the test with 
```python
xr.testing.assert_allclose(v_transport_theta_manual.isel(time=0), v_transport_theta.isel(time=0).rename('theta_on_v':'theta'))
```

**Comment by jdldeauna** [comment]
Got it, thanks!

**Review by andersy005 — APPROVED** [review]
Fantastic work, @jdldeauna! 

Quick question for @jbusecke, the notebook size is quite large `~85MB` due to the output cells: 

Screen Shot 2021-05-26 at 12 18 37 PM

Do you usually have these notebooks executed as part of the documentation builds on RTD? If so, would it be convenient to strip notebook output before committing the notebook?

**Comment by jbusecke** [comment]
> Do you usually have these notebooks executed as part of the documentation builds on RTD? If so, would it be convenient to strip notebook output before committing the notebook?

I think we used to, but RTD would usually struggle with the environments. Perhaps we could discuss this in another issue?

**Comment by jbusecke** [comment]
Thank you very much for significantly improving the documentation of xgcm, @jdldeauna.

**Comment by dcherian** [comment]
Nice work @jdldeauna

**Comment by jdldeauna** [comment]
Thank you for your patience! :)

**Pull request merged**